docs/operations/hook-sidecar.md

```diff
@@ -0,0 +1,162 @@
+information](https://libvirt.org/formatdomain.html#smbios-system-information) can be found in the [KubeVirt
+repo](https://github.com/kubevirt/kubevirt/tree/main/cmd/sidecars/smbios). Binary generated from this code, when
+available under `/usr/bin/ondefinedomain` in the sidecar-shim-image, is run right before VMI creation and the baseboard
+manufacturer value is modified to reflect what's provided in the `smbios.vm.kubevirt.io/baseBoardManufacturer`
+annotation in [VMI spec](https://github.com/kubevirt/kubevirt/blob/main/examples/vmi-with-sidecar-hook.yaml).
+
+### Shell or Python script
+
+If you pefer writing a shell or python script instead of a Go program, create a Kubernetes ConfigMap and use
+annotations to make sure the script is run before the VMI creation. The flow would be as below:
+
+1. Create a ConfigMap containing the shell or python script you want to run
+1. Create a VMI containing the annotation `hooks.kubevirt.io/hookSidecars` and mention the ConfigMap information in it.
+
+#### ConfigMap with shell script
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: my-config-map
+data:
+  my_script.sh: |
+    #!/bin/sh
+    tempFile=`mktemp --dry-run`
+    echo $4 > $tempFile
+    sed -i "s|<baseBoard></baseBoard>|<baseBoard><entry name='manufacturer'>Radical Edward</entry></baseBoard>|" $tempFile
+    cat $tempFile
+```
+
+#### ConfigMap with python script
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: my-config-map
+data:
+  my_script.sh: |
+    #!/usr/bin/env python
+
+    import xml.etree.ElementTree as ET
+    import sys
+
+    def main(s):
+        # write to a temporary file
+        f = open("/tmp/orig.xml", "w")
+        f.write(s)
+        f.close()
+
+        # parse xml from file
+        xml = ET.parse("/tmp/orig.xml")
+        # get the root element
+        root = xml.getroot()
+        # find the baseBoard element
+        baseBoard = root.find("sysinfo").find("baseBoard")
+
+        # prepare new element to be inserted into the xml definition
+        element = ET.Element("entry", {"name": "manufacturer"})
+        element.text = "Radical Edward"
+        # insert the element
+        baseBoard.insert(0, element)
+
+        # write to a new file
+        xml.write("/tmp/new.xml")
+        # print file contents to stdout
+        f = open("/tmp/new.xml")
+        print(f.read())
+        f.close()
+
+    if __name__ == "__main__":
+        main(sys.argv[4])
+```
+
+After creating one of the above ConfigMap, create the VMI using the manifest in [this
+example](https://github.com/kubevirt/kubevirt/blob/main/examples/vmi-with-sidecar-hook-configmap.yaml). Of importance
+here is the ConfigMap information stored in the annotations:
+
+```yaml
+annotations:
+  hooks.kubevirt.io/hookSidecars: '[{"args": ["--version", "v1alpha2"],
+    "image": "registry:5000/kubevirt/sidecar-shim:devel",
+    "configMap": {"name": "my-config-map", "key": "my_script.sh", "hookPath": "/usr/bin/onDefineDomain"}}]'
```

Comment:
The other day, I found [this example](https://github.com/kubernetes/kubernetes/pull/19758/commits/c8c82c1d8f8e143f9b571a731d7ff11d5101de20) as to how break down json annotations in a better format.

I just tested this with `examples/vmi-with-sidecar-hook.yaml`, the diff is:
```diff
-    hooks.kubevirt.io/hookSidecars: '[{"args": ["--version", "v1alpha2"], "image":
-      "registry:5000/kubevirt/example-hook-sidecar:devel"}]'
+    hooks.kubevirt.io/hookSidecars: >
+      [
+        {
+          "args": ["--version", "v1alpha2"],
+          "image": "registry:5000/kubevirt/example-hook-sidecar:devel"
+        }
+      ]
```

what do you think, about using this format instead?

Reply:
Wow! Amended.